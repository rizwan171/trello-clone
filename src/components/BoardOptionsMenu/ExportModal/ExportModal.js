import React, { useState } from "react";
import { useDispatch, useSelector } from "react-redux";

const ExportModal = ({ closeModal, handleExportList }) => {
  const lists = useSelector((state) => state.lists.value);
  const dispatch = useDispatch();
  const [selectedListId, setSelectedListId] = useState("");

  const handleChange = (e) => {
    setSelectedListId(e.target.value);
  };

  return (
    <div
      id="cardModal"
      className="z-50 min-h-full min-w-full flex fixed top-0 left-0 justify-center items-center bg-black bg-opacity-30"
    >
      <div className="relative px-4 w-full max-w-sm h-full md:h-auto mb-36">
        <div className="relative bg-trello-gray-100 rounded-lg shadow dark:bg-gray-700">
          <div className="flex p-6">
            <div className="">
              <select
                value={selectedListId}
                onChange={handleChange}
                className="form-select p-6 appearance-none w-md px-3 py-1.5 text-base font-normal text-gray-700 bg-clip-padding bg-no-repeat border border-solid border-gray-300 rounded transition ease-in-out m-0 focus:text-gray-700 focus:bg-white focus:border-blue-600 focus:outline-none"
              >
                <option selected value="" disabled>
                  Select List...
                </option>
                {lists.map((list) => {
                  return (
                    <option key={list.id} value={list.id}>
                      {list.title}
                    </option>
                  );
                })}
              </select>
            </div>
          </div>
          <div className="flex items-center px-6 pb-6">
            <button
              type="button"
              className="text-white bg-trello-green-100 hover:bg-trello-green-200 text-base rounded-lg px-5 py-2.5 text-center dark:bg-blue-600 dark:hover:bg-blue-700 dark:focus:ring-blue-800"
              onClick={() => handleExportList(selectedListId)}
            >
              Export
            </button>
            <button
              type="button"
              className="text-gray-600 bg-gray-300 text-base rounded-lg ml-2 px-5 py-2.5 text-center dark:bg-blue-600 dark:hover:bg-blue-700 dark:focus:ring-blue-800"
              onClick={() => closeModal()}
            >
              Cancel
            </button>
          </div>
        </div>
      </div>
    </div>
  );
};

export default ExportModal;
