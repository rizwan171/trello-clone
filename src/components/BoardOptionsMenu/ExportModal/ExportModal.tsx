import { ChangeEvent, useState } from "react";
import { useAppSelector } from "../../../app/hooks";
import { ExportModalProps } from "../../../types/components/ExportModalProps";

const ExportModal = ({ closeExportModal, handleExportList }: ExportModalProps): JSX.Element => {
  const lists = useAppSelector((state) => state.lists.value);
  const [selectedListId, setSelectedListId] = useState("");

  const handleChange = (e: ChangeEvent<HTMLSelectElement>) => {
    setSelectedListId(e.target.value);
  };

  return (
    <div
      id="cardModal"
      className="z-50 min-h-full min-w-full flex fixed top-0 left-0 justify-center items-center bg-black bg-opacity-30"
    >
      <div className="relative px-4 w-full max-w-sm h-full md:h-auto mb-36">
        <div className="relative bg-trello-gray-100 rounded-lg shadow dark:bg-gray-700">
          <div className="flex p-6">
            <div className="">
              <select
                value={selectedListId}
                onChange={handleChange}
                className="form-select p-6 appearance-none w-md px-3 py-1.5 text-base font-normal text-gray-700 bg-clip-padding bg-no-repeat border border-solid border-gray-300 rounded transition ease-in-out m-0 focus:text-gray-700 focus:bg-white focus:border-blue-600 focus:outline-none"
              >
                <option selected value="" disabled>
                  Select List...
                </option>
                {lists.map((list) => {
                  return (
                    <option key={list.id} value={list.id}>
                      {list.title}
                    </option>
                  );
                })}
              </select>
            </div>
          </div>
          <div className="flex items-center px-6 pb-6">
            <button
              type="button"
              className="text-white bg-trello-green-100 hover:bg-trello-green-200 text-base rounded-lg px-5 py-2.5 text-center dark:bg-blue-600 dark:hover:bg-blue-700 dark:focus:ring-blue-800"
              onClick={() => handleExportList(selectedListId)}
            >
              Export
            </button>
            <button
              type="button"
              className="text-gray-600 bg-gray-300 text-base rounded-lg ml-2 px-5 py-2.5 text-center dark:bg-blue-600 dark:hover:bg-blue-700 dark:focus:ring-blue-800"
              onClick={() => closeExportModal()}
            >
              Cancel
            </button>
          </div>
        </div>
      </div>
    </div>
  );
};

export default ExportModal;
