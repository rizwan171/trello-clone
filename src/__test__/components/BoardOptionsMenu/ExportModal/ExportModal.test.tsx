import { fireEvent, RenderResult } from "@testing-library/react";
import ExportModal from "../../../../components/BoardOptionsMenu/ExportModal/ExportModal";
import { renderWithProviders } from "../../../utils/test-utils";

describe("ExportModal", () => {
  const mockCloseExportModal = jest.fn();
  const mockHandleExportList = jest.fn();

  let component: RenderResult;
  let asFragment: DocumentFragment;

  beforeEach(() => {
    component = renderWithProviders(
      <ExportModal closeExportModal={mockCloseExportModal} handleExportList={mockHandleExportList} />,
      {
        preloadedState: {
          lists: {
            value: [
              { id: "1", title: "Test 1" },
              { id: "2", title: "Test 2" },
              { id: "3", title: "Test 3" },
              { id: "4", title: "Test 4" },
              { id: "5", title: "Test 5" },
            ],
          },
        },
      }
    );
    asFragment = component.asFragment();
  });

  it("should render successfully", () => {
    expect(asFragment).toMatchSnapshot();
  });

  it("should update selected list", () => {
    const selectOptions = component.getAllByRole("option") as HTMLOptionElement[];
    // no. of lists + 1 extra default option
    const expectedOptionValues = ["", "1", "2", "3", "4", "5"];
    const expectedOptionTexts = ["Select List...", "Test 1", "Test 2", "Test 3", "Test 4", "Test 5"];
    const actualOptionValues = selectOptions.map((option) => option.value);
    const actualOptionTexts = selectOptions.map((option) => option.text);
    expect(selectOptions.length).toEqual(6);
    expect(actualOptionValues).toStrictEqual(expectedOptionValues);
    expect(actualOptionTexts).toStrictEqual(expectedOptionTexts);

    const selector = component.getByRole("combobox") as HTMLSelectElement;
    expect(selector.value).toEqual("");

    fireEvent.select(selector, { target: { value: "1" } });
    expect(selector.value).toEqual("1");

    fireEvent.select(selector, { target: { value: "4" } });
    expect(selector.value).toEqual("4");
  });

  it("should call handleExportList", () => {
    const selector = component.getByRole("combobox") as HTMLSelectElement;
    fireEvent.select(selector, { target: { value: "1" } });
    expect(selector.value).toEqual("1");

    const exportButton = component.getByText("Export");
    fireEvent.click(exportButton);

    expect(mockHandleExportList).toHaveBeenCalled();
  });

  it("should call closeExportModal", () => {
    const closeButton = component.getByText("Cancel");
    fireEvent.click(closeButton);
    expect(mockCloseExportModal).toHaveBeenCalled();
  });
});
