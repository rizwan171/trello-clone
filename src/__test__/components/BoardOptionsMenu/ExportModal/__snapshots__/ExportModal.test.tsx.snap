// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ExportModal should render successfully 1`] = `
<DocumentFragment>
  <div
    class="z-50 min-h-full min-w-full flex fixed top-0 left-0 justify-center items-center bg-black bg-opacity-30"
    id="cardModal"
  >
    <div
      class="relative px-4 w-full max-w-sm h-full md:h-auto mb-36"
    >
      <div
        class="relative bg-trello-gray-100 rounded-lg shadow dark:bg-gray-700"
      >
        <div
          class="flex p-6"
        >
          <div>
            <select
              class="form-select p-6 appearance-none w-md px-3 py-1.5 text-base font-normal text-gray-700 bg-clip-padding bg-no-repeat border border-solid border-gray-300 rounded transition ease-in-out m-0 focus:text-gray-700 focus:bg-white focus:border-blue-600 focus:outline-none"
            >
              <option
                disabled=""
                value=""
              >
                Select List...
              </option>
              <option
                value="1"
              >
                Test 1
              </option>
              <option
                value="2"
              >
                Test 2
              </option>
              <option
                value="3"
              >
                Test 3
              </option>
              <option
                value="4"
              >
                Test 4
              </option>
              <option
                value="5"
              >
                Test 5
              </option>
            </select>
          </div>
        </div>
        <div
          class="flex items-center px-6 pb-6"
        >
          <button
            class="text-white bg-trello-green-100 hover:bg-trello-green-200 text-base rounded-lg px-5 py-2.5 text-center dark:bg-blue-600 dark:hover:bg-blue-700 dark:focus:ring-blue-800"
            type="button"
          >
            Export
          </button>
          <button
            class="text-gray-600 bg-gray-300 text-base rounded-lg ml-2 px-5 py-2.5 text-center dark:bg-blue-600 dark:hover:bg-blue-700 dark:focus:ring-blue-800"
            type="button"
          >
            Cancel
          </button>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
